GoogleDataTransport/GDTCORLibrary/GDTCORPlatform.m

```diff
@@ -330,21 +330,100 @@ - (instancetype)init {
+ */
++ (GDTCORBackgroundIdentifier)populateBgIDwithSemaphore:(dispatch_semaphore_t)semaphore {
+  __block GDTCORBackgroundIdentifier bgID = GDTCORBackgroundIdentifierInvalid;
+  if ([self getSemaphoreQueue]) {
+    dispatch_sync([self getSemaphoreQueue], ^{
+      NSMutableDictionary<NSNumber *, dispatch_semaphore_t> *bgIDToSemaphoreMap =
+          [self getBgIDToSemaphoreMap];
+      bgID = arc4random();
+      while (bgID == GDTCORBackgroundIdentifierInvalid ||
+             bgIDToSemaphoreMap[[NSNumber numberWithLong:bgID]]) {
+        bgID = arc4random();
+      }
+      bgIDToSemaphoreMap[[NSNumber numberWithLong:bgID]] = semaphore;
+    });
+  }
+  return bgID;
+}
+
+/** Returns the semaphore mapped be given bgID and removed the key value pair in
+ * NSMutableDictionary.
+ *
+ * @param bgID The unique NSUInteger as GDTCORBackgroundIdentifier.
+ * @return The semaphore mapped by given bgID.
+ */
++ (dispatch_semaphore_t)semaphoreForBgID:(GDTCORBackgroundIdentifier)bgID {
+  __block dispatch_semaphore_t semaphore;
+  if ([self getSemaphoreQueue]) {
+    dispatch_sync([self getSemaphoreQueue], ^{
+      NSMutableDictionary<NSNumber *, dispatch_semaphore_t> *bgIDToSemaphoreMap =
+          [self getBgIDToSemaphoreMap];
+      semaphore = bgIDToSemaphoreMap[[NSNumber numberWithLong:bgID]];
+      [bgIDToSemaphoreMap removeObjectForKey:[NSNumber numberWithLong:bgID]];
+    });
+  }
+  return semaphore;
+}
+
 - (GDTCORBackgroundIdentifier)beginBackgroundTaskWithName:(NSString *)name
                                         expirationHandler:(void (^)(void))handler {
+  __block GDTCORBackgroundIdentifier bgID = GDTCORBackgroundIdentifierInvalid;
 #if !TARGET_OS_WATCH
-  GDTCORBackgroundIdentifier bgID =
-      [[self sharedApplicationForBackgroundTask] beginBackgroundTaskWithName:name
-                                                           expirationHandler:handler];
+  bgID = [[self sharedApplicationForBackgroundTask] beginBackgroundTaskWithName:name
+                                                              expirationHandler:handler];
 #if !NDEBUG
   if (bgID != GDTCORBackgroundIdentifierInvalid) {
     GDTCORLogDebug("Creating background task with name:%@ bgID:%ld", name, (long)bgID);
   }
 #endif  // !NDEBUG
+#elif TARGET_OS_WATCH
+  [[self sharedNSProcessInfoForBackgroundTask]
+      performExpiringActivityWithReason:name
+                             usingBlock:^(BOOL expired) {
+                               dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
+                               bgID = [GDTCORApplication populateBgIDwithSemaphore:semaphore];
+                               if (expired) {
+                                 handler();
+                                 dispatch_semaphore_signal(semaphore);
+                               }
+                               dispatch_semaphore_wait(
+                                   semaphore, dispatch_time(DISPATCH_TIME_NOW, 30 * NSEC_PER_SEC));
```

Comment:
Any particular reasoning behind 30s?

Reply:
No idea. I thought it should be some explicit time expect FOREVER.

Reply:
I think every task we create should run very fast much less than 30s. Or should we change this waiting time less than 30s?

Reply:
Seems fine, I just wanted to know if there was reasoning. If we know about how long background tasks have to execute on watchOS before the expiration handler, we should set it to shorter than that.